device/pcie/pci_device.cpp

```diff
@@ -493,3 +499,83 @@ void PCIDevice::write_tlb_reg(uint32_t byte_addr, std::uint64_t value_lower, std
     }
     tt_driver_atomics::mfence(); // Otherwise subsequent WC loads move earlier than the above UC store to the TLB register.
 }
+
+
+bool PCIDevice::is_hardware_hung() {
+    volatile const void *addr = reinterpret_cast<const char *>(bar0_uc) + (get_architecture_implementation()->get_arc_reset_scratch_offset() + 6 * 4) - bar0_uc_offset;
+    std::uint32_t scratch_data = *reinterpret_cast<const volatile std::uint32_t*>(addr);
+
+    return (scratch_data == 0xffffffffu);
+}
+
+bool PCIDevice::reset_by_sysfs() {
+
+    const char *virtual_env = getenv("VIRTUAL_ENV");
+    if (virtual_env == nullptr)
+        return false;
+
+    std::string reset_helper_path = virtual_env;
+    reset_helper_path += "/bin/reset-helper";
+
+    std::string busid = std::to_string(info.pci_bus);
+
+    suspend_before_device_reset();
+
+    char *argv[3];
+    argv[0] = const_cast<char*>(reset_helper_path.c_str());
+    argv[1] = const_cast<char*>(busid.c_str());
+    argv[2] = nullptr;
+
+    pid_t reset_helper_pid;
+    if (posix_spawn(&reset_helper_pid, reset_helper_path.c_str(), nullptr, nullptr, argv, environ) != 0)
+        return false;
+
+    siginfo_t reset_helper_status;
+    if (waitid(P_PID, reset_helper_pid, &reset_helper_status, WEXITED) != 0)
+        return false;
+
+    if (reset_helper_status.si_status != 0)
+        return false;
+
+    resume_after_device_reset();
+
+    return true;
+}
+
+bool PCIDevice::reset_by_ioctl() {
+    struct tenstorrent_reset_device reset_device;
+    memset(&reset_device, 0, sizeof(reset_device));
+
+    reset_device.in.output_size_bytes = sizeof(reset_device.out);
+    reset_device.in.flags = 0;
+
+    if (ioctl(device_fd, TENSTORRENT_IOCTL_RESET_DEVICE, &reset_device) == -1) {
+        return false;
+    }
+
+    return (reset_device.out.result == 0);
+}
+
+bool PCIDevice::auto_reset_board() {
+    return ((reset_by_ioctl() || reset_by_sysfs()) && !is_hardware_hung());
+}
+
+void PCIDevice::detect_ffffffff_read(std::uint32_t data_read) {
+    if (data_read == 0xffffffffu && is_hardware_hung()) {
+        std::uint32_t scratch_data = *get_register_address<std::uint32_t>(read_checking_offset);
+
+        if (auto_reset_board()) {
+            throw std::runtime_error("Read 0xffffffff from PCIE: auto-reset succeeded.");
+        } else {
+            throw std::runtime_error("Read 0xffffffff from PCIE: you should reset the board.");
```

Comment:
What should the application do if it catches a `std::runtime_error`?  Unless it parses the `what()` string, the only reasonable thing to do is to terminate the program.

Either branch here is a fatal error.

Reply:
I think that is exactly the purpose of runtime_error. To tell the user this is fatal. 
Yes, both branches return runtime_error, as normal execution cannot continue after hang is detected. It is obvious from this code that the intention for auto reset feature was just to conveniently reset the card if it can, not try to continue normal execution.